overlord/hookstate/ctlcmd/kmod.go

```diff
@@ -0,0 +1,198 @@
+
+	kmodLoadModule   = kmod.LoadModule
+	kmodUnloadModule = kmod.UnloadModule
+)
+
+func init() {
+	addCommand("kmod", shortKmodHelp, longKmodHelp, func() command {
+		cmd := &kmodCommand{}
+		cmd.InsertCmd.kmod = cmd
+		cmd.RemoveCmd.kmod = cmd
+		return cmd
+	})
+}
+
+type kmodNopExecuteMixin struct {
+	kmod *kmodCommand
+}
+
+func (m *kmodNopExecuteMixin) Execute([]string) error {
+	// This is needed in order to implement the interface, but it's never
+	// called.
+	return nil
+}
+
+type kmodCommand struct {
+	baseCommand
+	kmodNopExecuteMixin
+	InsertCmd KModInsertCmd `command:"insert" description:"load a kernel module"`
+	RemoveCmd KModRemoveCmd `command:"remove" description:"unload a kernel module"`
+}
+
+type KModInsertCmd struct {
+	kmodNopExecuteMixin
+	Positional struct {
+		Module  string   `positional-arg-name:"<module>" required:"yes" description:"name of the kernel module to be loaded"`
+		Options []string `positional-arg-name:"<options>" description:"kernel module options"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+func (k *KModInsertCmd) Execute([]string) error {
+	context, err := k.kmod.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	attributes, err := kmodFindConnection(context, k.Positional.Module, k.Positional.Options)
+	if err != nil {
+		return fmt.Errorf("cannot load module %q: %v", k.Positional.Module, err)
+	}
+
+	if len(attributes) == 0 {
+		snapName := context.InstanceName()
+		return fmt.Errorf("snap %q lacks permissions to load the module %q", snapName, k.Positional.Module)
+	}
+
+	if err := kmodLoadModule(k.Positional.Module, k.Positional.Options); err != nil {
+		return fmt.Errorf("cannot load module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+type KModRemoveCmd struct {
+	kmodNopExecuteMixin
+	Positional struct {
+		Module string `positional-arg-name:"<module>" required:"yes" description:"name of the kernel module to be unloaded"`
+	} `positional-args:"yes" required:"yes"`
+	kmod *kmodCommand
+}
+
+func (k *KModRemoveCmd) Execute([]string) error {
+	context, err := k.kmod.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	attributes, err := kmodFindConnection(context, k.Positional.Module, []string{})
+	if err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	if len(attributes) == 0 {
+		snapName := context.InstanceName()
+		return fmt.Errorf("snap %q lacks permissions to unload the module %q", snapName, k.Positional.Module)
+	}
+
+	if err := kmodUnloadModule(k.Positional.Module); err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+// kmodMatchConnection checks whether the given kmod connection attributes give
+// the snap permission to execute the kmod command
+func kmodMatchConnection(attributes map[string]interface{}, moduleName string, moduleOptions []string) bool {
+	if attributes["load"].(string) != "dynamic" {
```

Comment:
the interfaces code suggest that load is optional, so maybe there should be an earlier check
```suggestion
	load, found := attributes["load"]
	if !found || load.(string) != "dynamic" {
		return false
	}
```

Reply:
Done